[ecxm02] iP

ecxm02 opened the pull request.

ecxm02 committed:
Original File (Renamed Class)

ecxm02 committed:
Week 2 (Level 0 IP, Rename Greet Exit)

ecxm02 committed:
Add Echo

Added simple reply functionality to output the input and end the program when a "bye" is detected

ecxm02 committed:
Add task and list

Add the ability to add tasks and list out all tasks

ecxm02 committed:
Add ability to Mark/Unmark

Add the ability to Mark and Unmark tasks. Made Task a class on its own, and changed the reply method to use switch cases instead.

ecxm02 committed:
Amend to follow coding standards

Removed unused library and added comments

ecxm02 committed:
Change programming style to OOP

Functionally the same as previous commit (tag - ACodingStandard), but followed OOP guidelines by splitting responsilbities of each class. (Still work in progress)

ecxm02 committed:
Add 3 new types of task (Todo, Deadline, Event)

Used inheritance to extend the class "Task" into 3 new types of tasks. Added specific variables that each type of task requires in the new classes. Used polymorphism and method overloading so that when the "TaskManager" class requests for getTaskType(), the appropriate task type is returned.

Improved parser so that it methodically splits the text input into the (Input Command, context of text, appropriate deadlines/from/to), allowing for better seperation into different variables (parsing method can be more efficient though)

ecxm02 committed:
Test script using IO redirection

Used the provided runtest bash file to check if the program's actual output is the same as the expected output. Provided in the commit is the script, input and expected text file

ecxm02 committed:
Improve Code Quality

Removed magic numbers and switched them to constants for better context, refactored some lines of text in ParseText to shorten each method length, ensured variables are consistent and made sense to user.

clj55 commented on src/main/java/ParseText.java:
```diff
@@ -21,27 +25,22 @@ private void textCleanup() {
         }
 
         else if (textInput.startsWith("deadline")) {
```
'else if' and 'else' should be on the same line as the previous }

clj55 commented on src/main/java/ParseText.java:
```diff
@@ -21,27 +25,22 @@ private void textCleanup() {
-            int slashIndex2 = textInput.indexOf("/", slashIndex + 1);
-            from = textInput.substring(slashIndex + 6, slashIndex2 - 1);
-            to = textInput.substring(slashIndex2 + 4);
+            taskName = textInput.substring(spaceIndex + Constants.EMPTY_PAD, slashIndex - Constants.EMPTY_PAD);
```
I like how you shorten it, makes it more readable :D

clj55 commented on src/main/java/ParseText.java:
```diff
@@ -21,27 +25,22 @@ private void textCleanup() {
+            taskName = textInput.substring(spaceIndex + Constants.EMPTY_PAD);
         }
 
         else {
```
here too, else should be on the same line as the previous }

ecxm02 replied:
Okay!

clj55 commented on src/main/java/Juke.java:
```diff
@@ -3,7 +3,8 @@
     private static boolean isCompleted = false; //a master flag so that main knows when to end the program
 
-    public static void endProgram() { //sets the isCompleted flag to true so that the loop ends, also prints end message
+    //sets the isCompleted flag to true so that the loop ends, also prints end message
```
I think comments about the method (indented on same level as endProgram like how you have done) should describe what the method does overall. If you want to explain how the code works maybe put it inside the the method (under end Program and indented to match isCompleted

ecxm02 replied:
Okays

clj55 reviewed (commented):
Overall clean code yay :D 
Side note: you can use the string.split function instead of getting the index then doing substring

dylankhoo commented on src/main/java/Juke.java:
```diff
@@ -0,0 +1,23 @@
+
+        while (!isCompleted) {
+            text = input.nextLine();
+            ReplyText current = new ReplyText(text);
```
Seeing that you are not using the "current" variable elsewhere, rather than creating new instances of ReplyText, could it be better to just call a method from ReplyText here?

ecxm02 replied:
Yea I will change some classes into static and call their methods instead

dylankhoo commented on src/main/java/ReplyText.java:
```diff
@@ -0,0 +1,46 @@
+
+    public ReplyText(String textInput) {
+        this.textInput = textInput;
+        ParseText cleaner = new ParseText(textInput);
```
A more intuitive variable name here could perhaps be "cleanedText"

ecxm02 replied:
Okays

dylankhoo commented on src/main/java/ParseText.java:
```diff
@@ -0,0 +1,74 @@
+        textCleanup();
+    }
+
+    private void textCleanup() {
+        if (textInput.startsWith("mark") || textInput.startsWith("unmark")) {
+            String[] words = textInput.split(" ");
+            String taskNumber = words[1];
+            taskIndex = Integer.parseInt(taskNumber) - 1;
+            inputCommand = words[0];
+        }
+
+        else if (textInput.startsWith("deadline")) {
+            inputCommand = textInput.substring(0, spaceIndex);
+            taskName = textInput.substring(spaceIndex + Constants.EMPTY_PAD, slashIndex - Constants.EMPTY_PAD);
+            deadline = textInput.substring(slashIndex + Constants.BY_PAD);
+        }
+
+        else if (textInput.startsWith("event")) {
+            inputCommand = textInput.substring(0, spaceIndex);
+            taskName = textInput.substring(spaceIndex + Constants.EMPTY_PAD, slashIndex - Constants.EMPTY_PAD);
+            int slashIndex2 = textInput.indexOf("/", slashIndex + Constants.EMPTY_PAD);
+            from = textInput.substring(slashIndex + Constants.FROM_PAD, slashIndex2 - Constants.EMPTY_PAD);
+            to = textInput.substring(slashIndex2 + Constants.TO_PAD);
+        }
+
+        else if (textInput.startsWith("todo")) {
+            inputCommand = textInput.substring(0, spaceIndex);
+            taskName = textInput.substring(spaceIndex + Constants.EMPTY_PAD);
+        }
+
+        else {
+            inputCommand = textInput;
+        }
+    }
```
This is quite a long method, perhaps you could consider breaking it into smaller methods? You could have individual methods (e,.g. cleanUpMark) for each command and call them accordingly based on textInput.

ecxm02 replied:
Alright

dylankhoo commented on src/main/java/ReplyText.java:
```diff
@@ -0,0 +1,46 @@
+public class ReplyText {
+    private static TaskManager taskList = new TaskManager();
```
Rather than creating the TaskManager instance here, could it be perhaps more intuitive to create the instance in Juke instead?

ecxm02 replied:
Ok! Will amend

dylankhoo reviewed (commented):
Overall I like how you extracted your code into different classes and kept Juke clean.

However, I do find that in some places, it may be more appropriate to call static methods rather than constructing instances of new classes. 

I have left some comments on where I felt that this was the case.

ecxm02 committed:
Clean code based on PR Reviews

Renamed classes ReplyText to InputHandler and ParseText to Parser for better clarity, and made them static with private constructors so no instances can be called. (It didnt make sense for them to have different instances)
Formatted some if else statements to match coding standards

Initialised taskList in main as it made more sense instead of in inputHandler
Broke down textCleaner method in Parser to allow for higher level abstraction

ecxm02 committed:
Add resetVariable method

Added a method to reset all the variables in Parser class. Helps with detecting invalid inputs by cleaning up after every input

ecxm02 committed:
Add error handling for invalid input

Added ability to detect invalid inputs or incomplete inputs for the tasks. Moved all error checking to Parser Class, combined some methods to clean up and improve logical flow.

ecxm02 committed:
Merge pull request #1 from ecxm02/branch-Level-5

branch-Level-5

ecxm02 committed:
Merge branch 'master' of https://github.com/ecxm02/ip

ecxm02 committed:
Add custom exception ParserExecption

Handles exceptions by throwing a custom exception called ParserException whenever an invalid command syntax is provided. (Will add more specific exceptions in the future)

ecxm02 committed:
Split classes into separate packages

Added 3 new packages, main, task and input (respective classes are inside these packages)

ecxm02 committed:
Merge pull request #2 from ecxm02/branch-A-Packages

branch-A-Packages

ecxm02 committed:
Made exception its own class

Will add more exceptions in the future

ecxm02 committed:
Switch from Array to Arraylist

Switched to ArrayList from Array to simplify task storage

ecxm02 committed:
Add delete

Added the ability to delete tasks

ecxm02 committed:
Add ability to save

Saves on exit, will now change to save on task add

ecxm02 committed:
Modify Save

Now saves every time a new task is added. Improved formatting of list and comands in general

ecxm02 committed:
Merge pull request #3 from ecxm02/branch-Level-6

Combine branch-Level-6 to master

ecxm02 committed:
Merge branch 'master' into branch-Level-7

ecxm02 committed:
Merge pull request #4 from ecxm02/branch-Level-7

Merge branch-Level-7 to master

ecxm02 committed:
Fix Merge

Minor issues caused by Merging level 6 and level 7 branches into main. Fixed in this commit

ecxm02 committed:
Setup JAR

Created an executable JAR file and uploaded the first release on GitHub

ecxm02 committed:
Revert "Setup JAR"

This reverts commit 489d5e60ca8dd83546df5e710b0430e5111c7302.

ecxm02 committed:
Remove Save File

Shouldn't be in repo

ecxm02 committed:
Revert back to A-Jar after accidental revert

Made an error

ecxm02 committed:
Merge branch 'master' of https://github.com/ecxm02/ip

ecxm02 committed:
Revert back to A-Jar

ecxm02 committed:
Reapply "Setup JAR"

This reverts commit edd2a18973e1b20e47d4cfc13699363b1ca93b85.

ecxm02 committed:
Fixed all the insane messy commits and reverts

sevenseasofbri commented on src/main/java/juke/input/InputHandler.java:
```diff
@@ -0,0 +1,54 @@
+                Parser.getDeadline(), Parser.getFrom(), Parser.getTo());
+    }
+
+    //uses a switch to execute appropriate methods for the input
```
Implementation details like this can be left out of the comments. This may not be providing any useful input to the reader.

sevenseasofbri commented on src/main/java/juke/input/InputHandler.java:
```diff
@@ -0,0 +1,54 @@
+    public static void respondToText(TaskManager taskManager, String inputCommand, String taskName, int taskIndex,
+                                     String deadline, String from, String to) {
+
+        switch (inputCommand) {
```
Good job keeping a consistent level of abstraction.

sevenseasofbri commented on src/main/java/juke/input/Parser.java:
```diff
@@ -0,0 +1,154 @@
+import juke.main.Constants;
+
+public class Parser {
+    private static String textInput;
+    private static String inputCommand = "invalid";
+    private static String taskName;
+    private static String taskNumber;
+    private static int taskIndex = -1;
+    private static int spaceIndex = -1;
+    private static int slashIndex = -1;
+    private static int slashIndex2 = -1;
+    private static String deadline;
+    private static String from;
+    private static String to;
```
If any of these are not going to change in the duration of the program, you can change them to constants `static final` in `SCREAMING_SNAKE_CASE`

sevenseasofbri commented on src/main/java/juke/input/Parser.java:
```diff
@@ -0,0 +1,154 @@
+    private Parser() {
+    }
+
+    public static void cleanText(String text) {
```
This function is slightly >30 LOC. Consider chunking it into multiple methods.

sevenseasofbri commented on src/main/java/juke/input/Parser.java:
```diff
@@ -0,0 +1,154 @@
+        slashIndex = textInput.indexOf("/", spaceIndex + Constants.EMPTY_PAD);
+        slashIndex2 = textInput.indexOf("/", slashIndex + Constants.EMPTY_PAD);
+
+        if (textInput.startsWith("mark") || textInput.startsWith("unmark")) {
+            try {
+                cleanUpMark();
+            } catch (Exception e) {
+                inputCommand = "invalid";
+            }
+        } else if (textInput.startsWith("todo")) {
+            try {
+                cleanUpTodo();
+            } catch (Exception e) {
+                inputCommand = "invalid";
+            }
+        } else if (textInput.startsWith("deadline")) {
+            try {
+                cleanUpDeadline();
+            } catch (Exception e) {
+                inputCommand = "invalid";
+            }
+        } else if (textInput.startsWith("event")) {
+            try {
+                cleanUpEvent();
+            } catch (Exception e) {
+                inputCommand = "invalid";
+            }
+        } else if (textInput.startsWith("delete")) {
+            try {
+                cleanUpDelete();
+            } catch (Exception e) {
+                inputCommand = "invalid";
+            }
```
You can refactor these try catch blocks into their own methods.

sevenseasofbri commented on src/main/java/juke/persistence/FileSaver.java:
```diff
@@ -0,0 +1,60 @@
+    public static void writeToFile(Task task) {
+        ensureDirectoryExists();
+
+        try (FileWriter writer = new FileWriter((Constants.SAVE_FILE_NAME), true)) {
+            String taskType = task.getTaskType();
+            switch (taskType) {
+            case "T":
+                writer.write("T" + "|" + task.isDone() + "|" + task.getTaskName() + System.lineSeparator());
+                break;
+            case "D":
+                Deadline deadlineTask = (Deadline) task;
+                writer.write("D" + "|" + task.isDone() + "|" + deadlineTask.getTaskName() + "|"
+                        + deadlineTask.getDeadline() + System.lineSeparator());
+                break;
+            case "E":
+                Event eventTask = (Event) task;
+                writer.write("E" + "|" + task.isDone() + "|" + eventTask.getTaskName() + "|"
+                        + eventTask.getFrom() + "|" + eventTask.getTo() + System.lineSeparator());
+                break;
```
Consider extracting these complex expressions into thir own methods.

sevenseasofbri commented on src/main/java/juke/task/TaskManager.java:
```diff
@@ -0,0 +1,109 @@
+    public TaskManager() {
+    }
+
+    public void listTask() { //prints out the tasks and marks them appropriately
```
Consider using JavaDoc format to describe methods instead.

sevenseasofbri reviewed (commented):
Great job so far! Some segments of the code have deep nesting (>=3 levels of indentation, e.g.. `readFromFile` from FileLoader) so you might want to try and look resloving that.

ecxm02 committed:
More OOP

Added a new UI class, made TaskManager static since we did not need multiple instances of it

ecxm02 committed:
Added Date

Added standard date format to deadline, to and from. Inteprets it as a actual date instead of just a string. Added even more exceptions and better error handling

ecxm02 committed:
Merge pull request #5 from ecxm02/branch-Level-8

Added Date

ecxm02 committed:
Add Regex

Added regex to check syntax of inputs, improved error handling significantly

ecxm02 committed:
Add Find

Added ability to find tasks through the new keyword, find

ecxm02 committed:
Update Exception and Constant

Updated some exceptions and constants to standardise things

ecxm02 committed:
Merge pull request #6 from ecxm02/branch-Level-9

Branch level 9

ecxm02 committed:
Add Javadoc Comments

Added Javadoc comments in all clases except exceptions

ecxm02 committed:
Merge pull request #7 from ecxm02/branch-A-JavaDoc

Add Javadoc Comments

ecxm02 committed:
Updated ReadME

ecxm02 committed:
Update README.md

ecxm02 committed:
Update ReadME

Updated both ReadMe

ecxm02 committed:
Bug Fixes

Bug Fix on delete method
